helix-term/src/commands/lsp.rs

```diff
@@ -108,6 +113,83 @@ fn sym_picker(
     .truncate_start(false)
 }
 
+fn diag_picker(
+    cx: &Context,
+    diagnostics: BTreeMap<lsp::Url, Vec<lsp::Diagnostic>>,
+    current_path: Option<lsp::Url>,
+    offset_encoding: OffsetEncoding,
+) -> FilePicker<(lsp::Url, lsp::Diagnostic)> {
+    // TODO: drop current_path comparison and instead use workspace: bool flag?
+
+    // flatten the map to a vec of (uri, diag) pairs
+    let mut flat_diag = Vec::new();
+    for (u, diags) in diagnostics {
```

Comment:
add a `flat_diag.reserve(diags.len())` before this line. Replacing the imperative loop with `flat_map` could work too.

Reply:
I played around with it but bidn't get it to work with `flat_map`I had a version with some nested folds but it looked too complicated, so I left the straight-forward nested loops.